Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h

```diff
@@ -40,44 +40,43 @@ namespace WebKit {
 
 class WebCacheStorageProvider;
 
-class WebCacheStorageConnection final : public CanMakeWeakPtr<WebCacheStorageConnection>, public WebCore::CacheStorageConnection {
+class WebCacheStorageConnection final : public WebCore::CacheStorageConnection {
 public:
-    static Ref<WebCacheStorageConnection> create(WebCacheStorageProvider& provider) { return adoptRef(*new WebCacheStorageConnection(provider)); }
+    static Ref<WebCacheStorageConnection> create() { return adoptRef(*new WebCacheStorageConnection); }
 
     ~WebCacheStorageConnection();
 
     void didReceiveMessage(IPC::Connection&, IPC::Decoder&);
     void networkProcessConnectionClosed();
 
 private:
-    explicit WebCacheStorageConnection(WebCacheStorageProvider&);
+    WebCacheStorageConnection();
 
     template<typename P> struct PromiseConverter;
 
-    IPC::Connection& connection();
-
     // WebCore::CacheStorageConnection
     Ref<OpenPromise> open(const WebCore::ClientOrigin&, const String& cacheName) final;
     Ref<RemovePromise> remove(WebCore::DOMCacheIdentifier) final;
-
     Ref<RetrieveCachesPromise> retrieveCaches(const WebCore::ClientOrigin&, uint64_t)  final;
-
     Ref<RetrieveRecordsPromise> retrieveRecords(WebCore::DOMCacheIdentifier, WebCore::RetrieveRecordsOptions&&)  final;
     Ref<BatchPromise> batchDeleteOperation(WebCore::DOMCacheIdentifier, const WebCore::ResourceRequest&, WebCore::CacheQueryOptions&&)  final;
     Ref<BatchPromise> batchPutOperation(WebCore::DOMCacheIdentifier, Vector<WebCore::DOMCacheEngine::CrossThreadRecord>&&)  final;
-
     void reference(WebCore::DOMCacheIdentifier) final;
     void dereference(WebCore::DOMCacheIdentifier) final;
     void lockStorage(const WebCore::ClientOrigin&) final;
     void unlockStorage(const WebCore::ClientOrigin&) final;
-
     Ref<CompletionPromise> clearMemoryRepresentation(const WebCore::ClientOrigin&) final;
     Ref<EngineRepresentationPromise> engineRepresentation() final;
     void updateQuotaBasedOnSpaceUsage(const WebCore::ClientOrigin&) final;
 
-    WebCacheStorageProvider& m_provider;
-    HashCountedSet<WebCore::DOMCacheIdentifier> m_connectedIdentifierCounters;
-    HashCountedSet<WebCore::ClientOrigin> m_clientOriginLockRequestCounters;
+    Ref<IPC::Connection> connection();
+
+    Lock m_connectionLock;
+    RefPtr<IPC::Connection> m_connection;
+    Lock m_connectedIdentifierCountersLock;
+    HashCountedSet<WebCore::DOMCacheIdentifier> m_connectedIdentifierCounters WTF_GUARDED_BY_LOCK(m_connectedIdentifierCountersLock);
+    Lock m_clientOriginLockRequestCountersLock;
```

Comment:
Can we just have one lock (e.g. `m_connectionLock`) and make all members protected by it since all operations require it now? Then we don't need to ensure the ordering of the locks

Reply:
Good point, it does not bring much to have 3 locks.

Reply:
Done